app/utility/base_knowledge_svc.py

```diff
@@ -71,7 +72,21 @@ async def _get_meta_facts(self, meta_fact=None, agent=None, group=None):
     async def _get_fact_origin(self, fact):
         # Retrieve the specific origin of a fact. If it was learned in the current operation, parse through links to
         # identify the host it was discovered on.
-        raise NotImplementedError
+        factOrigin = 'None'
+        try:
+            if fact.links:
+                factOrigin = str(fact.links[0].host)
+            else:
+                factOrigin = str(fact.source)
+        except AttributeError:
+            foundFacts = await self._get_facts(fact)
+            if foundFacts and len(foundFacts) >= 1:
+                if foundFacts[0].links:
+                    factOrigin = str(foundFacts[0].links[0].host)
```

Comment:
This is being super pedantic, but this line (and the other on 87) both seem to be pretty darn similar to 78 and 80. Is there any way we could simplify these to reduce the duplication?